pkg/utils/kubernetes/health/health.go

```diff
@@ -0,0 +1,262 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package health
+
+import (
+	"fmt"
+	gardenv1alpha1 "github.com/gardener/machine-controller-manager/pkg/apis/machine/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func requiredConditionMissing(conditionType string) error {
+	return fmt.Errorf("condition %q is missing", conditionType)
+}
+
+func checkConditionState(conditionType string, expected, actual, reason, message string) error {
+	if expected != actual {
+		return fmt.Errorf("condition %q has invalid status %s (expected %s) due to %s: %s",
+			conditionType, expected, actual, reason, message)
+	}
+	return nil
+}
+
+func getDeploymentCondition(conditions []appsv1.DeploymentCondition, conditionType appsv1.DeploymentConditionType) *appsv1.DeploymentCondition {
+	for _, condition := range conditions {
+		if condition.Type == conditionType {
+			return &condition
+		}
+	}
+	return nil
+}
+
+func getMachineDeploymentCondition(conditions []gardenv1alpha1.MachineDeploymentCondition, conditionType gardenv1alpha1.MachineDeploymentConditionType) *gardenv1alpha1.MachineDeploymentCondition {
+	for _, condition := range conditions {
+		if condition.Type == conditionType {
+			return &condition
+		}
+	}
+	return nil
+}
+
+func getNodeCondition(conditions []corev1.NodeCondition, conditionType corev1.NodeConditionType) *corev1.NodeCondition {
+	for _, condition := range conditions {
+		if condition.Type == conditionType {
+			return &condition
+		}
+	}
+	return nil
+}
+
+var (
+	trueDeploymentConditionTypes = []appsv1.DeploymentConditionType{
+		appsv1.DeploymentAvailable,
+		appsv1.DeploymentProgressing,
+	}
+
+	falseDeploymentConditionTypes = []appsv1.DeploymentConditionType{
+		appsv1.DeploymentReplicaFailure,
+	}
+)
+
+// CheckDeployment checks whether the given Deployment is healthy.
+// A deployment is considered healthy if the controller observed its current revision and
+// if the number of updated replicas is equal to the number of replicas.
+func CheckDeployment(deployment *appsv1.Deployment) error {
+	if deployment.Status.ObservedGeneration < deployment.Generation {
+		return fmt.Errorf("observed generation outdated (%d/%d)", deployment.Status.ObservedGeneration, deployment.Generation)
```

Comment:
Should we have a more user-friendly message here? Something like "Deployment will be updated" or "Deployment update processing" or something similar?

Reply:
(same for the other occurrences)

Reply:
Check `checkRequiredDeployments` or `checkDeployments`: The user message is being made more friendly by prefixing the affected k8s object's name.
Also, by my current design, the health checks should only print the issues with the object but not the objects name itself.